File: core/core-web/src/test/java/ch/difty/scipamato/core/web/paper/search/SearchOrderSelectorPanelInViewModeTest.java
package ch.difty.scipamato.core.web.paper.search;

import static org.mockito.ArgumentMatchers.isA;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import org.junit.jupiter.api.Test;

import ch.difty.scipamato.common.web.Mode;
import ch.difty.scipamato.core.entity.search.SearchOrder;

class SearchOrderSelectorPanelInViewModeTest extends SearchOrderSelectorPanelTest {

    @Override
    Mode getMode() {
        return Mode.VIEW;
    }

    @Test
    void withGlobalSearchOrders_withSameOwner_globalCheckBox_disabled() {
        when(searchOrderMock.getName()).thenReturn(VALID_NAME);
        when(searchOrderMock.getOwner()).thenReturn(OWNER_ID);
        getTester().startComponentInPage(makePanel());

        getTester().assertDisabled(PANEL_ID + ":form:global");
    }

    @Test
    void withGlobalSearchOrders_withOtherOwner_globalCheckBox_disabled() {
        when(searchOrderMock.getName()).thenReturn(VALID_NAME);
        when(searchOrderMock.getOwner()).thenReturn(OWNER_ID + 1);
        getTester().startComponentInPage(makePanel());

        getTester().assertDisabled(PANEL_ID + ":form:global");
    }

    @Test
    void changingName_forGlobalNotSelfOwnedSearch_disablesName() {
        when(searchOrderMock.getName()).thenReturn(VALID_NAME);
        when(searchOrderMock.isGlobal()).thenReturn(true);
        when(searchOrderMock.getOwner()).thenReturn(OWNER_ID + 1);
        assertNameDisabled();
    }

    private void assertNameDisabled() {
        getTester().startComponentInPage(makePanel());
        getTester().assertDisabled(PANEL_ID + ":form:name");
    }

    @Test
    void changingName_forGlobalSelfOwnedSearch_disablesName() {
        when(searchOrderMock.getName()).thenReturn(VALID_NAME);
        when(searchOrderMock.isGlobal()).thenReturn(true);
        when(searchOrderMock.getOwner()).thenReturn(OWNER_ID);
        assertNameDisabled();
    }

    @Test
    void changingName_forNotGlobalNotSelfOwnedSearch_disablesName() {
        when(searchOrderMock.getName()).thenReturn(VALID_NAME);
        when(searchOrderMock.isGlobal()).thenReturn(false);
        when(searchOrderMock.getOwner()).thenReturn(OWNER_ID + 1);
        assertNameDisabled();
    }

    @Test
    void changingName_forNotGlobalSelfOwnedSearch_doesAddTargetAndSaves() {
        when(searchOrderMock.getName()).thenReturn(VALID_NAME);
        when(searchOrderMock.isGlobal()).thenReturn(false);
        when(searchOrderMock.getOwner()).thenReturn(OWNER_ID);
        getTester().startComponentInPage(makePanel());

        getTester().executeAjaxEvent(PANEL_ID + ":form:name", "change");

        verify(searchOrderServiceMock).saveOrUpdate(isA(SearchOrder.class));
    }
}

File: core/core-web/src/test/kotlin/ch/difty/scipamato/core/web/paper/search/SearchOrderSelectorPanelInViewModeTest.kt
package ch.difty.scipamato.core.web.paper.search

import ch.difty.scipamato.common.web.Mode
import io.mockk.verify
import org.junit.jupiter.api.Test

internal class SearchOrderSelectorPanelInViewModeTest : SearchOrderSelectorPanelTest() {

    override val mode: Mode
        get() = Mode.VIEW

    @Test
    fun withGlobalSearchOrders_withSameOwner_globalCheckBox_disabled() {
        tester.startComponentInPage(makePanel())
        tester.assertDisabled("$PANEL_ID:form:global")
    }

    @Test
    fun withGlobalSearchOrders_withOtherOwner_globalCheckBox_disabled() {
        searchOrder.apply { owner = OWNER_ID + 1 }
        tester.startComponentInPage(makePanel())
        tester.assertDisabled("$PANEL_ID:form:global")
    }

    @Test
    fun changingName_forGlobalNotSelfOwnedSearch_disablesName() {
        searchOrder.apply {
            isGlobal = true
            owner = OWNER_ID + 1
        }
        assertNameDisabled()
    }

    private fun assertNameDisabled() {
        tester.startComponentInPage(makePanel())
        tester.assertDisabled("$PANEL_ID:form:name")
    }

    @Test
    fun changingName_forGlobalSelfOwnedSearch_disablesName() {
        searchOrder.apply { isGlobal = true }
        assertNameDisabled()
    }

    @Test
    fun changingName_forNotGlobalNotSelfOwnedSearch_disablesName() {
        searchOrder.apply {
            isGlobal = true
            owner = OWNER_ID + 1
        }
        assertNameDisabled()
    }

    @Test
    fun changingName_forNotGlobalSelfOwnedSearch_doesAddTargetAndSaves() {
        searchOrder.apply { isGlobal = false }
        tester.startComponentInPage(makePanel())
        tester.executeAjaxEvent("$PANEL_ID:form:name", "change")
        verify { searchOrderServiceMock.saveOrUpdate(any()) }
    }
}
